tfs/writer.py

```diff
@@ -234,35 +306,115 @@ def _dtype_to_id_string(type_: type) -> str:
+
+
+def _dtype_to_string_format_id(type_: type) -> str:
+    """
+    Return the string-formatter type-identifier for the provided dtype.
+    Of special note are here "b" for boolean and "c" for complex numbers.
+
+    Args:
+        type_ (type): an instance of the built-in type (in this package, one of
+            ``numpy`` or ``pandas`` types) to get the formatter for.
+
+    Returns:
+        str: the formatter type-identifier.
     """
+
     if type_ is None:
-        return f"{colsize}"
-    if pdtypes.is_integer_dtype(type_) or pdtypes.is_bool_dtype(type_):
-        return f"{colsize}d"
+        return ""
+    if pdtypes.is_integer_dtype(type_):
+        return "d"
+    if pdtypes.is_bool_dtype(type_):
+        return "b"  # can only be used with TfsStringFormatter
     if pdtypes.is_float_dtype(type_):
-        return f"{colsize}.{colsize - len('-0.e-000')}g"
+        return "g"
     if pdtypes.is_string_dtype(type_):
-        return f"{colsize}s"
-    errmsg = f"Provided type '{type_}' could not be identified as either a bool, int, float or string dtype"
+        return "s"
+    if pdtypes.is_complex_dtype(type_):
+        return "c"  # can only be used with TfsStringFormatter
+
+    errmsg = (
+        f"Provided type '{type_}' could not be identified as either a bool, int, float, complex or string dtype"
+    )
     raise TypeError(errmsg)
+
+
+class TfsStringFormatter(string.Formatter):
+    """Formatter class to be called for proper formatting of TFS strings."""
+
+    def format_field(self, value, format_spec):
+        if format_spec.endswith("b"):
+            return self._format_boolean(value, format_spec)
+
+        elif format_spec.endswith("c"):
+            return self._format_complex(value, format_spec)
+
+        if format_spec.endswith("s"):
+            return self._format_string(value, format_spec)
+
+        return super().format_field(value, format_spec)
+
+    def _format_boolean(self, value, format_spec: str):
+        """
+        Special case for booleans, as from their written version
+        will always be either 'true' or 'false' (lowercase) as
+        done by MAD-NG.
+        """
+        bool_str = str(bool(value)).lower()
+        return super().format_field(bool_str, f"{format_spec[:-1]}s")
+
+    def _format_complex(self, value, format_spec: str):
+        """
+        Special case for complex numbers, MAD-X and MAD-NG use 'i' for
+        the imaginary part and not 'j', so we have to replace that in.
+        """
+        return super().format_field(value, f"{format_spec[:-1]}g").replace("j", "i")
+
+    def _format_string(self, value, format_spec: str) -> str:
+        """
+        Special case as we need to ensure that strings are enclosed
+        in either " or ' quotes.
+        """
+        try:
+            if not value.startswith(('"', "'")):
+                value = f'"{value}"'
+        except AttributeError:
+            pass
+        return super().format_field(value, format_spec)
```

Comment:
**FEATURE**

String representation is now handled here.

It expects the value and the format_spec, which we determine ahead of time in `_dtype_to_string_format_id` based on the dtype for the value.

Default handling from `string.Formatter` is done, and special cases are written for boolean and complex dtypes.

Note that this is not slower than the previous implementation for all previously accepted dtypes, actually a little faster :) Big thanks to @JoschD for the implementation.